Fix kubeconfig-sync issues

jakolehm opened the pull request:
Fixes regressions from #2567

- `CatalogPusher` was not syncing changes to item properties
- `KubeconfigSync` 
  - always created a new `Cluster` object
  - did not create stable `id` for a cluster
  - called `cluster.init()` (for no reason?)
- `ClusterManager` logic was completely broken

jakolehm committed:
fix kubeconfig-sync issues

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

jakolehm commented on src/main/catalog-pusher.ts:
```diff
@@ -14,8 +14,8 @@ export class CatalogPusher {
 
-    disposers.push(reaction(() => this.catalog.items, (items) => {
-      broadcastMessage("catalog:items", toJS(items, { recurseEverything: true }));
+    disposers.push(reaction(() => toJS(this.catalog.items, { recurseEverything: true }), (items) => {
```
Otherwise changes to catalog items won't trigger reaction.

Nokel81 replied:
Okay... this breaks my mental model of how reactions work.

jakolehm commented on src/main/catalog-sources/kubeconfig-sync.ts:
```diff
@@ -24,7 +24,7 @@ export class KubeconfigSyncManager extends Singleton {
 
   @action
-  startSync(port: number): void {
+  startSync(): void {
```
Passing port multiple layers down to just for `cluster.init()` felt wrong. Init is anyway called from `ClusterManager`.

Nokel81 replied:
So I have noticed that all of `Cluster.init()` can be hoisted into `Cluster.constructor`.

Nokel81 replied:
Also I specifically didn't want to rely on `ClusterManager`. Wouldn't it be possible to query this for a cluster ID instead?

jakolehm commented on src/main/catalog-sources/kubeconfig-sync.ts:
```diff
@@ -156,14 +156,13 @@ export function computeDiff(contents: string, source: RootSource, port: number,
             throw new Error("Cluster constructor failed, see above error");
           }
 
-          cluster.init(port);
```
This is done via `ClusterManager` (for now).

jakolehm commented on src/main/catalog-sources/kubeconfig-sync.ts:
```diff
@@ -156,14 +156,13 @@ export function computeDiff(contents: string, source: RootSource, port: number,
         // add new clusters to the source
         try {
-          const cluster = new Cluster({ ...model, id: uuid.v4() });
+          const clusterId = createHash("md5").update(`${filePath}:${contextName}`).digest("hex");
```
We need stable id for catalog<->cluster relation.

Nokel81 replied:
Sounds good, I already knew we would need this

jakolehm commented on src/main/catalog-sources/kubeconfig-sync.ts:
```diff
@@ -156,14 +156,13 @@ export function computeDiff(contents: string, source: RootSource, port: number,
         try {
-          const cluster = new Cluster({ ...model, id: uuid.v4() });
+          const clusterId = createHash("md5").update(`${filePath}:${contextName}`).digest("hex");
+          const cluster = ClusterStore.getInstance().getById(clusterId) || new Cluster({ ...model, id: clusterId});
```
Uses saved cluster for base if it exists. Otherwise we don't see any saved changes on catalog items.

Nokel81 replied:
If `clusterId` already exists we shouldn't even be in this code path...

jakolehm commented on src/main/cluster-manager.ts:
```diff
@@ -1,24 +1,21 @@
   constructor(public readonly port: number) {
     super();
 
-    catalogEntityRegistry.addObservableSource("lens:kubernetes-clusters", this.catalogSource);
```
Was probably forgotten during the original PR. We don't need this now that the source is `KubeconfigSync`.

Nokel81 replied:
See #2663, there is a lot that I knew I didn't to do. Was planning to do this in the future.

Nokel81 committed:
Cleanup Cluster, ClusterManger, et al. (#2698)

- Removed `getFreePort` as its use is always a race condition. Change
  all uses of it to retrive the port after listening

- Added `getPortFrom` as a helper function to read a port from a stream

- Remove `Cluster.ownerRef` as it is outdated and no longer needed for 5.0

- Remove `Cluster.enabled`, no longer needed because of above

- Removed `Cluster.init`, moved its contents into `Cluster.constructor`
  as nothing in that function is asyncronous. Currently only being run
  on `main` as a stop gap until `renderer` gets its own version of
  `Cluster`

- Refactored `LensProxy` so as to prevent `pty.node` (a NodeJS
  extension) being included in `webpack.extension.ts`'s run

- Removed the passing around of the proxy port as that can now be
  accessed from an instance of `LensProxy`

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm committed:
Merge branch 'master' into fix/kubeconfig-sync-issues-part-1

jakolehm committed:
tweak add cluster page

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

jakolehm committed:
show delete only for internally stored clusters

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

jakolehm committed:
fix tests

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

jakolehm committed:
fix tests

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

jakolehm committed:
fix tests

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

jakolehm committed:
fix tests

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

jakolehm committed:
fix tests

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

Nokel81 committed:
purge ContextHandler's cache on disconnect

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
remove adding minikube, should be auto detected

Signed-off-by: Sebastian Malton <sebastian@malton.name>
